src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlOptionMarkupExtensionTransformer.cs

```diff
@@ -0,0 +1,405 @@
+            VisitList(Branches, visitor);
+            DefaultNode = (IXamlAstValueNode)DefaultNode?.Visit(visitor);
+        }
+
+        public IXamlType GetReturnType()
+        {
+            var types = Branches.Select(b => b.Value.Type);
+            if (DefaultNode?.Type is { } type)
+            {
+                types = types.Append(type);
+            }
+            return types.Select(t => t.GetClrType()).ToArray().GetCommonBaseClass();
+        }
+    }
+
+    internal sealed class OptionsMarkupExtensionBranch : XamlAstNode
+    {
+        public OptionsMarkupExtensionBranch(IXamlAstValueNode option, IXamlAstValueNode value, IXamlMethod conditionMethod) : base(value)
+        {
+            Option = option;
+            Value = value;
+            ConditionMethod = conditionMethod;
+        }
+
+        public IXamlAstValueNode Option { get; set; }
+        public IXamlAstValueNode Value { get; set; }
+        public IXamlMethod ConditionMethod { get; }
+
+        public bool HasContext => ConditionMethod.Parameters.Count > 1;
+
+        public override void VisitChildren(IXamlAstVisitor visitor)
+        {
+            Option = (IXamlAstValueNode)Option.Visit(visitor);
+            Value = (IXamlAstValueNode)Value.Visit(visitor);
+        }
+    }
+
+    internal sealed class OptionsMarkupExtensionMethod : IXamlCustomEmitMethodWithContext<IXamlILEmitter, XamlILNodeEmitResult>
+    {
+        public OptionsMarkupExtensionMethod(
+            OptionsMarkupExtensionNodesContainer extensionNodeContainer,
+            IXamlType declaringType,
+            IXamlType contextParameter)
+        {
+            ExtensionNodeContainer = extensionNodeContainer;
+            DeclaringType = declaringType;
+            Parameters = extensionNodeContainer.Branches.Any(c => c.HasContext) ?
+                new[] { contextParameter } :
+                Array.Empty<IXamlType>();
+        }
+
+        public OptionsMarkupExtensionNodesContainer ExtensionNodeContainer { get; }
+
+        public string Name => "ProvideValue";
+        public bool IsPublic => true;
+        public bool IsStatic => false;
+        public IXamlType ReturnType => ExtensionNodeContainer.GetReturnType();
+        public IReadOnlyList<IXamlType> Parameters { get; }
+        public IXamlType DeclaringType { get; }
+        public IXamlMethod MakeGenericMethod(IReadOnlyList<IXamlType> typeArguments) => throw new NotImplementedException();
+        public IReadOnlyList<IXamlCustomAttribute> CustomAttributes => Array.Empty<IXamlCustomAttribute>();
+
+        public void EmitCall(XamlEmitContext<IXamlILEmitter, XamlILNodeEmitResult> context, IXamlILEmitter codeGen)
+        {
+            // At this point this extension will be called from MarkupExtensionEmitter.
+            // Since it's a "fake" method, we share stack and locals with parent method.
+            // Real ProvideValue method would pop 2 parameters from the stack and return one. This method should do the same.
```

Comment:
I think we have some custom methods that can skip parameters, not sure if they are supported by the markup extension emitter